## 1. Introduction

## 2. Data Acquisition

In [3]:
import requests
import pandas as pd
import json
import time
import os

In [9]:
import numpy as np
from pathlib import Path

data_path = Path.cwd().parent / "data"

work_bank_df = pd.read_csv(
    "https://ourworldindata.org/grapher/world-bank-income-groups.csv"
    "?v=1&csvType=full&useColumnShortNames=false",
    storage_options={"User-Agent": "Our World In Data data fetch/1.0"}
)

print("World Bank shape:", work_bank_df.shape)
print(work_bank_df.head())

WB_TO_ACLED = {
    "Congo":        "Republic of Congo",       # was "Congo, Rep."
    "Czechia":      "Czech Republic",
    "Eswatini":     "eSwatini",
    "Congo, Dem. Rep.":  "Democratic Republic of Congo",
    "Côte d'Ivoire":     "Ivory Coast",
    "Korea, Rep.":       "South Korea",
    "Korea, Dem. People's Rep.": "North Korea",
    "Lao PDR":           "Laos",
    "Syrian Arab Republic": "Syria",
    "Yemen, Rep.":       "Yemen",
    "Venezuela, RB":     "Venezuela",
    "Iran, Islamic Rep.": "Iran",
    "Egypt, Arab Rep.":  "Egypt",
    "Kyrgyz Republic":   "Kyrgyzstan",
    "Turkiye":           "Turkey",
}

work_bank_df["Entity"] = work_bank_df["Entity"].replace(WB_TO_ACLED)

INCOME_COL = "World Bank's income classification"

# Keep only the most recent classification per country
wb_latest = (
    work_bank_df
    .sort_values("Year", ascending=False)
    .drop_duplicates(subset="Entity")
)

global_south = wb_latest.loc[
    wb_latest[INCOME_COL] != 'High-income countries',
    'Entity'
].unique().tolist()

western = wb_latest.loc[
    wb_latest[INCOME_COL] == 'High-income countries',
    'Entity'
].unique().tolist()

# Manual overrides for geopolitically misclassified countries
OVERRIDES_WESTERN = ["Ukraine"]
OVERRIDES_GLOBAL_SOUTH = [
    "Saudi Arabia", "Qatar", "Kuwait",
    "Bahrain", "Oman", "United Arab Emirates"
]

for c in OVERRIDES_WESTERN:
    if c not in western:
        western.append(c)
    if c in global_south:
        global_south.remove(c)

for c in OVERRIDES_GLOBAL_SOUTH:
    if c not in global_south:
        global_south.append(c)
    if c in western:
        western.remove(c)

print(f"\nGlobal South countries: {len(global_south)}")
print(f"Western countries:      {len(western)}")

fatalities = pd.read_csv(
    f"{data_path}/number_of_reported_fatalities_by_country-year_as-of-03Apr2026.csv",
    skiprows=1
)

fatalities = fatalities.loc[:, ~fatalities.columns.str.contains('^Unnamed')]

print("\nFatalities shape:", fatalities.shape)
print(fatalities.head())

# Filter to study window 2015-2024 only
# ACLED has data from 1997 but coverage is sparse before 2015.
# We also exclude 2025 and 2026 which are outside our Guardian date range.
fatalities = fatalities[
    (fatalities['YEAR'] >= 2015) &
    (fatalities['YEAR'] <= 2024)
]

print(f"\nAfter date filter: {fatalities.shape}")
print(fatalities['YEAR'].value_counts().sort_index())

# =============================================================================
# ADD REGION CLASSIFICATION TO FATALITIES
# =============================================================================

fatalities["region_group"] = fatalities["COUNTRY"].apply(
    lambda x: "Western" if x in western else "Global South"
)

print("\nRegion group distribution:")
print(fatalities["region_group"].value_counts())

# Sanity check — countries not matched to either list
unmatched = fatalities[
    ~fatalities["COUNTRY"].isin(western + global_south)
]["COUNTRY"].unique()
print(f"\nUnmatched countries: {len(unmatched)}")
if len(unmatched) > 0:
    print(sorted(unmatched))

# =============================================================================
# SAVE CLEAN VERSION
# =============================================================================

fatalities.to_csv(f"{data_path}/cleaned_fatalities.csv", index=False)
print(f"\nSaved cleaned_fatalities.csv — shape: {fatalities.shape}")

World Bank shape: (7953, 4)
        Entity Code  Year World Bank's income classification
0  Afghanistan  AFG  1987               Low-income countries
1  Afghanistan  AFG  1988               Low-income countries
2  Afghanistan  AFG  1989               Low-income countries
3  Afghanistan  AFG  1990               Low-income countries
4  Afghanistan  AFG  1991               Low-income countries

Global South countries: 141
Western countries:      85

Fatalities shape: (2943, 3)
       COUNTRY  YEAR  FATALITIES
0  Afghanistan  2017       36360
1  Afghanistan  2018       42991
2  Afghanistan  2019       41419
3  Afghanistan  2020       31037
4  Afghanistan  2021       42425

After date filter: (1634, 3)
YEAR
2015     61
2016     74
2017     77
2018    161
2019    161
2020    212
2021    245
2022    215
2023    213
2024    215
Name: count, dtype: int64

Region group distribution:
region_group
Global South    1221
Western          413
Name: count, dtype: int64

Unmatched countries: 33
['Akroti

In [10]:
with open("../keys.json") as f:
    keys = json.load(f)

API_KEYS = [
    keys["guardian"]["api_key_1"],
    keys["guardian"]["api_key_2"],
    keys["guardian"]["api_key_3"]
]


In [11]:
acled = pd.read_csv("../data/cleaned_fatalities.csv")

print("=== ACLED Dataset ===")
print(f"Shape: {acled.shape}")
print(f"Columns: {list(acled.columns)}")
print(f"Countries: {acled['COUNTRY'].nunique()} unique")
print(f"Year range: {acled['YEAR'].min()} – {acled['YEAR'].max()}")
acled.head()

=== ACLED Dataset ===
Shape: (1634, 4)
Columns: ['COUNTRY', 'YEAR', 'FATALITIES', 'region_group']
Countries: 245 unique
Year range: 2015 – 2024


,COUNTRY,YEAR,FATALITIES,region_group
0,Afghanistan,2017,36360,Global South
1,Afghanistan,2018,42991,Global South
2,Afghanistan,2019,41419,Global South
3,Afghanistan,2020,31037,Global South
4,Afghanistan,2021,42425,Global South


In [12]:
FROM_DATE = "2015-01-01"
TO_DATE   = "2024-12-31"

# --- Automatic conversion: lowercase and replace spaces with hyphens ---
acled_countries = acled["COUNTRY"].unique().tolist()

def name_to_slug(name):
    return name.lower().strip().replace(" ", "-").replace("'", "").replace(",", "")

auto_slugs = {country: name_to_slug(country) for country in acled_countries}

In [13]:
SLUG_CORRECTIONS = {
    "Democratic Republic of Congo": "democratic-republic-of-congo",
    "Republic of Congo":            "republic-of-congo",
    "United States":                "us-news",
    "Myanmar":                      "myanmar",
    "Ivory Coast":                  "ivory-coast",
    "Burkina Faso":                 "burkina-faso",
    "South Sudan":                  "south-sudan",
    "Central African Republic":     "central-african-republic",
    "Bosnia-Herzegovina":           "bosnia-and-herzegovina",
    "North Korea":                  "north-korea",
    "South Korea":                  "south-korea",
    "Saudi Arabia":                 "saudi-arabia",
    "Sri Lanka":                    "sri-lanka",
    "Papua New Guinea":             "papua-new-guinea",
}

In [14]:
COUNTRY_TO_SLUG = {**auto_slugs, **SLUG_CORRECTIONS}

# Final list of (country_name, guardian_slug) pairs
COUNTRY_LIST = [(country, COUNTRY_TO_SLUG[country]) for country in acled_countries]

print(f"\nTotal countries to query: {len(COUNTRY_LIST)}")
print("\nSample mappings:")
for name, slug in COUNTRY_LIST[:5]:
    print(f"  {name} -> world/{slug}")


Total countries to query: 245

Sample mappings:
  Afghanistan -> world/afghanistan
  Akrotiri and Dhekelia -> world/akrotiri-and-dhekelia
  Albania -> world/albania
  Algeria -> world/algeria
  American Samoa -> world/american-samoa


In [6]:
import os
if os.path.exists("../data/guardian_raw.csv"):
    os.remove("../data/guardian_raw.csv")
    print("Deleted old guardian_raw.csv — ready for fresh collection")

Deleted old guardian_raw.csv — ready for fresh collection


In [15]:
BASE_URL    = "https://content.guardianapis.com/search"
current_key = 0  

def fetch_articles(country_name, slug):
    """
    Fetches all Guardian articles for a given country between FROM_DATE
    and TO_DATE. Paginates automatically and rotates API keys on rate
    limiting.

    For USA, UK, and Australia the Guardian uses dedicated top-level
    sections rather than world/ tags. These are detected by the "section:"
    prefix in the slug and queried using the section parameter instead of tag.

    Parameters:
        country_name : str — original ACLED country name, stored in output
        slug         : str — either "section:sectionid" or a world/ tag slug

    Returns a list of article dicts.
    """
    global current_key
    articles = []
    page = 1

    # Detect whether to query by section or by tag
    if slug.startswith("section:"):
        query_param = {"section": slug.replace("section:", "")}
    else:
        query_param = {"tag": f"world/{slug}"}

    while True:
        params = {
            **query_param,
            "from-date":   FROM_DATE,
            "to-date":     TO_DATE,
            "show-fields": "wordcount,sectionName,byline",
            "page-size":   200,
            "page":        page,
            "api-key":     API_KEYS[current_key]
        }

        r = requests.get(BASE_URL, params=params)
        time.sleep(0.1)

        if r.status_code == 429:
            current_key = (current_key + 1) % len(API_KEYS)
            print(f"  Rate limited — switching to key {current_key}")
            continue

        if r.status_code != 200:
            print(f"  Error {r.status_code} for {slug} page {page} — skipping")
            break

        data    = r.json()["response"]
        results = data.get("results", [])

        if not results:
            break

        for a in results:
            articles.append({
                "country":   country_name,
                "date":      a["webPublicationDate"],
                "section":   a["sectionName"],
                "wordcount": a.get("fields", {}).get("wordcount", None),
                "title":     a["webTitle"],
                "pillar":    a.get("pillarName", None),
                "byline": a.get("fields", {}).get("byline", None),
            })

        if page >= data["pages"]:
            break

        page += 1

    return articles


In [16]:
# --- Run collection (skips if file already exists) ---
RAW_PATH = "../data/guardian_raw.csv"

if os.path.exists(RAW_PATH) and os.path.getsize(RAW_PATH) > 0:
    print("Guardian raw data already collected — loading from file.")
    guardian_raw = pd.read_csv(RAW_PATH)
else:
    all_articles = []
    for country_name, slug in COUNTRY_LIST:
        print(f"Fetching: {country_name}...")
        articles = fetch_articles(country_name, slug)
        all_articles.extend(articles)
        print(f"  -> {len(articles)} articles")

    guardian_raw = pd.DataFrame(all_articles)
    guardian_raw.to_csv(RAW_PATH, index=False)
    print(f"\nSaved {len(guardian_raw)} total articles to {RAW_PATH}")

print(f"\nShape: {guardian_raw.shape}")
print(f"Countries with articles: {guardian_raw['country'].nunique()}")
print(f"Date range: {guardian_raw['date'].min()} to {guardian_raw['date'].max()}")
guardian_raw.head()

Fetching: Afghanistan...
  -> 2656 articles
Fetching: Akrotiri and Dhekelia...
  -> 0 articles
Fetching: Albania...
  -> 162 articles
Fetching: Algeria...
  -> 130 articles
Fetching: American Samoa...
  -> 0 articles
Fetching: Andorra...
  -> 5 articles
Fetching: Angola...
  -> 117 articles
Fetching: Anguilla...
  -> 0 articles
Fetching: Antarctica...
  -> 434 articles
Fetching: Antigua and Barbuda...
  -> 46 articles
Fetching: Argentina...
  -> 736 articles
Fetching: Armenia...
  -> 185 articles
Fetching: Aruba...
  -> 0 articles
Fetching: Australia...
  -> 0 articles
Fetching: Austria...
  -> 623 articles
Fetching: Azerbaijan...
  -> 292 articles
Fetching: Bahamas...
  -> 121 articles
Fetching: Bahrain...
  -> 267 articles
Fetching: Bailiwick of Guernsey...
  -> 0 articles
Fetching: Bailiwick of Jersey...
  -> 0 articles
Fetching: Bangladesh...
  -> 699 articles
Fetching: Barbados...
  -> 99 articles
Fetching: Belarus...
  -> 465 articles
Fetching: Belgium...
  -> 1015 articles
Fetch

,country,date,section,wordcount,title,pillar,byline
0,Afghanistan,2024-12-31T18:41:27Z,UK news,495,Special forces troops face prosecution over al...,News,Kiran Stacey Political correspondent
1,Afghanistan,2024-12-29T17:09:11Z,Media,471,David Page obituary,News,William Crawley
2,Afghanistan,2024-12-23T17:00:21Z,US news,2008,‘Never underestimate what neighbors can do for...,News,David Wallis
3,Afghanistan,2024-12-06T17:50:37Z,Global development,446,Taliban move to ban women training as nurses a...,News,Ruchi Kumar and Zahra Joya for Rukhshana Media
4,Afghanistan,2024-12-06T10:40:29Z,Sport,471,Top Afghan cricketers urge Taliban to reverse ...,Sport,Jack Snape


In [17]:
def fetch_articles_slow(country_name, slug):
    """Same as fetch_articles but with longer sleep for large sections."""
    global current_key
    articles = []
    page = 1

    if slug.startswith("section:"):
        query_param = {"section": slug.replace("section:", "")}
    else:
        query_param = {"tag": f"world/{slug}"}

    while True:
        params = {
            **query_param,
            "from-date":   FROM_DATE,
            "to-date":     TO_DATE,
            "show-fields": "wordcount,sectionName,byline",
            "page-size":   200,
            "page":        page,
            "api-key":     API_KEYS[current_key]
        }

        try:
            r = requests.get(BASE_URL, params=params, timeout=30)
            time.sleep(0.5)  # slower — reduces stalling risk
        except requests.exceptions.Timeout:
            print(f"  Timeout on page {page} — retrying...")
            time.sleep(5)
            continue

        if r.status_code == 429:
            current_key = (current_key + 1) % len(API_KEYS)
            print(f"  Rate limited — switching to key {current_key}")
            time.sleep(2)
            continue

        if r.status_code != 200:
            print(f"  Error {r.status_code} on page {page} — stopping")
            break

        data    = r.json()["response"]
        results = data.get("results", [])
        if not results:
            break

        for a in results:
            articles.append({
                "country":   country_name,
                "date":      a["webPublicationDate"],
                "section":   a["sectionName"],
                "wordcount": a.get("fields", {}).get("wordcount", None),
                "title":     a["webTitle"],
                "pillar":    a.get("pillarName", None),
                "byline": a.get("fields", {}).get("byline", None),
            })

        if page % 50 == 0:
            print(f"  ...page {page}, {len(articles)} articles so far")

        if page >= data["pages"]:
            break

        page += 1

    return articles

In [18]:
aus_articles = fetch_articles_slow("Australia", "section:australia-news")
print(f"Australia: {len(aus_articles)} articles")

  ...page 50, 10000 articles so far
  ...page 100, 20000 articles so far
  ...page 150, 30000 articles so far
  Error 400 on page 191 — stopping
Australia: 38000 articles


In [19]:
existing = pd.read_csv("../data/guardian_raw.csv")
aus_df = pd.DataFrame(aus_articles)
guardian_raw = pd.concat([existing, aus_df], ignore_index=True)
guardian_raw.to_csv("../data/guardian_raw.csv", index=False)
print(f"Total after Australia: {len(guardian_raw)} articles")

Total after Australia: 201282 articles


In [20]:
uk_articles = fetch_articles_slow("United Kingdom", "section:uk-news")
print(f"UK: {len(uk_articles)} articles")

  ...page 50, 10000 articles so far
  ...page 100, 20000 articles so far
  ...page 150, 30000 articles so far
UK: 37297 articles


In [21]:
existing = pd.read_csv("../data/guardian_raw.csv")
uk_df = pd.DataFrame(uk_articles)
guardian_raw = pd.concat([existing, uk_df], ignore_index=True)
guardian_raw.to_csv("../data/guardian_raw.csv", index=False)
print(f"Total after UK: {len(guardian_raw)} articles")

Total after UK: 238579 articles


In [22]:
us_articles = fetch_articles_slow("United States", "section:us-news")
print(f"US: {len(us_articles)} articles")

  ...page 50, 10000 articles so far
  ...page 100, 20000 articles so far
  Rate limited — switching to key 1
  ...page 150, 30000 articles so far
  Error 400 on page 191 — stopping
US: 38000 articles


In [23]:
existing = pd.read_csv("../data/guardian_raw.csv")
us_df = pd.DataFrame(us_articles)
guardian_raw = pd.concat([existing, us_df], ignore_index=True)
guardian_raw.to_csv("../data/guardian_raw.csv", index=False)
print(f"Total after USA: {len(guardian_raw)} articles")

Total after USA: 276579 articles


## World Bank Income Classification

In [59]:
OWID_URL = (
    "https://ourworldindata.org/grapher/world-bank-income-groups.csv"
    "?v=1&csvType=full&useColumnShortNames=false"
)

wb_raw = pd.read_csv(
    OWID_URL,
    storage_options={"User-Agent": "Our World In Data data fetch/1.0"}
)

print("=== World Bank Income Classification ===")
print(f"Shape: {wb_raw.shape}")
print(f"Columns: {list(wb_raw.columns)}")
print(f"Years available: {wb_raw['Year'].min()} – {wb_raw['Year'].max()}")

# Keep only the most recent year — we use a single static classification
# per country rather than a time-varying one, which is sufficient for
# our binary Global South / Global North label.
wb_latest = (
    wb_raw
    .sort_values("Year", ascending=False)
    .drop_duplicates(subset="Entity")
    .rename(columns={"Entity": "country", "Year": "wb_year"})
)

# Save locally so the notebook is reproducible without an internet connection
wb_latest.to_csv("../data/worldbank_income.csv", index=False)
print(f"\nSaved to data/worldbank_income.csv")
print(f"Countries classified: {wb_latest['country'].nunique()}")
wb_latest.head()


=== World Bank Income Classification ===
Shape: (7953, 4)
Columns: ['Entity', 'Code', 'Year', "World Bank's income classification"]
Years available: 1987 – 2024

Saved to data/worldbank_income.csv
Countries classified: 226


,country,Code,wb_year,World Bank's income classification
4148,Liberia,LBR,2024,Low-income countries
7914,Zambia,ZMB,2024,Lower-middle-income countries
37,Afghanistan,AFG,2024,Low-income countries
4110,Lesotho,LSO,2024,Lower-middle-income countries
4186,Libya,LBY,2024,Upper-middle-income countries


In [23]:
INCOME_COL = [c for c in wb_latest.columns if "income" in c.lower()][0]
print(f"Using income column: '{INCOME_COL}'")

# Apply binary label
wb_latest["region_group"] = wb_latest[INCOME_COL].apply(
    lambda x: "Global North" if str(x).strip() == "High-income countries" else "Global South"
)

# --- Manual overrides for edge cases ---
# Some countries are misclassified by income group alone for our purposes.
# We document each override explicitly with a reason.
OVERRIDES = {
    # Geopolitically Global North despite upper-middle income classification
    "Ukraine":      "Global North",
    # High income but geopolitically and geographically Global South
    "Saudi Arabia": "Global South",
    "Qatar":        "Global South",
    "Kuwait":       "Global South",
    "Bahrain":      "Global South",
    "Oman":         "Global South",
    "United Arab Emirates": "Global South",
}

for country, label in OVERRIDES.items():
    mask = wb_latest["country"] == country
    if mask.any():
        wb_latest.loc[mask, "region_group"] = label
        print(f"  Override: {country} -> {label}")

# --- Build the two lists ---
# We classify all World Bank countries here without filtering to ACLED yet.
# Name harmonisation between World Bank and ACLED happens in Section 3.
# Filtering to conflict-affected countries only will be applied post-merge.

global_north_countries = sorted(
    wb_latest[wb_latest["region_group"] == "Global North"]["country"].tolist()
)

global_south_countries = sorted(
    wb_latest[wb_latest["region_group"] == "Global South"]["country"].tolist()
)

print(f"Global North countries in ACLED:      {len(global_north_countries)}")
print(f"Global South countries in ACLED: {len(global_south_countries)}")
print(f"Global North sample:      {global_north_countries[:5]}")
print(f"Global South sample: {global_south_countries[:5]}")

# Save the classification back to CSV for use in Section 3
wb_latest.to_csv("../data/worldbank_income.csv", index=False)
print("\nUpdated worldbank_income.csv with region_group column")

Using income column: 'World Bank's income classification'
  Override: Ukraine -> Global North
  Override: Saudi Arabia -> Global South
  Override: Qatar -> Global South
  Override: Kuwait -> Global South
  Override: Bahrain -> Global South
  Override: Oman -> Global South
  Override: United Arab Emirates -> Global South
Global North countries in ACLED:      85
Global South countries in ACLED: 141
Global North sample:      ['American Samoa', 'Andorra', 'Antigua and Barbuda', 'Aruba', 'Australia']
Global South sample: ['Afghanistan', 'Albania', 'Algeria', 'Angola', 'Argentina']

Updated worldbank_income.csv with region_group column


# 3. Preparation of the Data for Analysis

# IDA for guardian_raw file

## 1. Basic structure

In [25]:
print(guardian_raw.shape)
print(guardian_raw.dtypes)
print(guardian_raw.head())

(276579, 7)
country      object
date         object
section      object
wordcount    object
title        object
pillar       object
byline       object
dtype: object
       country                  date             section wordcount  \
0  Afghanistan  2024-12-31T18:41:27Z             UK news       495   
1  Afghanistan  2024-12-29T17:09:11Z               Media       471   
2  Afghanistan  2024-12-23T17:00:21Z             US news      2008   
3  Afghanistan  2024-12-06T17:50:37Z  Global development       446   
4  Afghanistan  2024-12-06T10:40:29Z               Sport       471   

                                               title pillar  \
0  Special forces troops face prosecution over al...   News   
1                                David Page obituary   News   
2  ‘Never underestimate what neighbors can do for...   News   
3  Taliban move to ban women training as nurses a...   News   
4  Top Afghan cricketers urge Taliban to reverse ...  Sport   

                                  

## 2. Missing values

In [26]:
print(guardian_raw.isnull().sum())
print(f"Wordcount missing: {guardian_raw['wordcount'].isnull().sum()} ({guardian_raw['wordcount'].isnull().mean()*100:.1f}%)")

country         0
date            0
section         0
wordcount       0
title           0
pillar       1101
byline       2118
dtype: int64
Wordcount missing: 0 (0.0%)


1101 pillar's missing is fine as less than 0.5% is missing (1101/276000) and 2118 is also fine as less than 0.8% missing (2118/276000)

## 3. Duplicates

In [28]:
print(f"Duplicate rows: {guardian_raw.duplicated().sum()}")
print(f"Duplicate titles: {guardian_raw.duplicated(subset='title').sum()}")

Duplicate rows: 0
Duplicate titles: 50500


The duplicates is fine as guardian uses generic title names such as "Ukraine war - Latest updates" or "News roundup"

## 4. Wordcount type check — it comes in as string, needs converting

In [29]:
print(guardian_raw['wordcount'].dtype)
print(guardian_raw['wordcount'].sample(10))

object
76591      889
194887     779
228817     760
55960      370
95254      391
66517     1030
4113       378
191185     687
106603     224
181414     783
Name: wordcount, dtype: object


## 5. Section distribution — are there unexpected sections?

In [30]:
print(guardian_raw['section'].value_counts().head(20))

section
World news            84890
US news               43716
Australia news        42085
UK news               41032
Opinion               13245
Global development     8103
Business               5973
Environment            5564
Politics               4502
Film                   3848
Books                  2325
Art and design         1915
Media                  1656
Music                  1564
Technology             1482
Sport                  1481
News                   1479
Cities                 1442
Football               1392
Science                1067
Name: count, dtype: int64


## 6. Date format check

In [31]:
print(guardian_raw['date'].sample(5))

77764     2019-12-12T00:31:03Z
230657    2017-01-14T12:54:50Z
248614    2023-06-13T13:02:12Z
90246     2024-10-29T11:26:09Z
209514    2022-12-14T18:10:34Z
Name: date, dtype: object


## 7. Articles per country — any countries with suspiciously low counts?

In [32]:
print(guardian_raw['country'].value_counts().tail(20))

country
Dominica                            26
Seychelles                          24
eSwatini                            23
Reunion                             23
Equatorial Guinea                   22
Togo                                22
Djibouti                            21
Saint Vincent and the Grenadines    20
Saint Kitts and Nevis               19
Turks and Caicos Islands            17
Suriname                            17
Cape Verde                          13
Guinea-Bissau                       13
Mayotte                             11
Montserrat                          10
San Marino                           9
Liechtenstein                        7
Andorra                              5
Comoros                              5
Sao Tome and Principe                5
Name: count, dtype: int64


## 8. Year distribution — any gaps?

In [33]:
guardian_raw['year'] = pd.to_datetime(guardian_raw['date']).dt.year
print(guardian_raw['year'].value_counts().sort_index())

year
2015    24888
2016    21280
2017    24875
2018    25527
2019    24802
2020    26265
2021    26355
2022    35129
2023    33595
2024    33863
Name: count, dtype: int64


A clear jump in articles written in 2022 (35128 articles) from the previous year (26355 articles) certainly in part motivated by the war in Ukraine.

# IDA for ACLED file (cleaned_fatalities)

## 1. Basic structure

In [34]:
print(acled.shape)
print(acled.dtypes)
print(acled.head())

(1634, 4)
COUNTRY         object
YEAR             int64
FATALITIES       int64
region_group    object
dtype: object
       COUNTRY  YEAR  FATALITIES  region_group
0  Afghanistan  2017       36360  Global South
1  Afghanistan  2018       42991  Global South
2  Afghanistan  2019       41419  Global South
3  Afghanistan  2020       31037  Global South
4  Afghanistan  2021       42425  Global South


## 2. Missing values

In [35]:
print(acled.isnull().sum())

COUNTRY         0
YEAR            0
FATALITIES      0
region_group    0
dtype: int64


## 3. Year range

In [63]:
print(acled['YEAR'].value_counts().sort_index())

YEAR
2015     61
2016     74
2017     77
2018    161
2019    161
2020    212
2021    245
2022    215
2023    213
2024    215
Name: count, dtype: int64


## 4. Fatalities distribution — any zeroes or negatives?

In [65]:
print(acled['FATALITIES'].describe())
print(f"Zero fatality rows: {(acled['FATALITIES'] == 0).sum()}")

count     1634.000000
mean       991.755202
std       4549.874372
min          0.000000
25%          0.000000
50%          4.000000
75%        150.000000
max      73573.000000
Name: FATALITIES, dtype: float64
Zero fatality rows: 566


## 5. Duplicates

In [66]:
print(f"Duplicate country-year pairs: {acled.duplicated(subset=['COUNTRY','YEAR']).sum()}")

Duplicate country-year pairs: 0


# IDA for World Bank income data 

## 1. Basic structure

In [69]:
wb_latest = wb_latest.reset_index(drop=True)
print(wb_latest.shape)
print(wb_latest[['country', INCOME_COL]].head(20))

(226, 4)
          country World Bank's income classification
0         Liberia               Low-income countries
1          Zambia      Lower-middle-income countries
2     Afghanistan               Low-income countries
3         Lesotho      Lower-middle-income countries
4           Libya      Upper-middle-income countries
5         Lebanon      Lower-middle-income countries
6   Liechtenstein              High-income countries
7          Latvia              High-income countries
8            Laos      Lower-middle-income countries
9       Lithuania              High-income countries
10     Kyrgyzstan      Lower-middle-income countries
11     Luxembourg              High-income countries
12         Kuwait              High-income countries
13          Macao              High-income countries
14         Kosovo      Upper-middle-income countries
15       Kiribati      Lower-middle-income countries
16     Madagascar               Low-income countries
17          Kenya      Lower-middle-i

## 2. How many in each group?

In [73]:
print(wb_latest[INCOME_COL].value_counts())

World Bank's income classification
High-income countries            90
Upper-middle-income countries    59
Lower-middle-income countries    51
Low-income countries             26
Name: count, dtype: int64


## 3. Any nulls in income column?

In [74]:
print(wb_latest[INCOME_COL].isnull().sum())

0


## 4. Countries in ACLED missing from World Bank classification

In [80]:
acled_names = set(acled['COUNTRY'].unique())
wb_names = set(wb_latest['Entity'].unique())
print("In ACLED but not World Bank:")
print(sorted(acled_names - wb_names))

In ACLED but not World Bank:
['Akrotiri and Dhekelia', 'Anguilla', 'Antarctica', 'Bailiwick of Guernsey', 'Bailiwick of Jersey', 'British Indian Ocean Territory', 'Caribbean Netherlands', 'Christmas Island', 'Cocos (Keeling) Islands', 'Cook Islands', 'Falkland Islands', 'French Southern and Antarctic Lands', 'Guadeloupe', 'Heard Island and McDonald Islands', 'Ivory Coast', 'Martinique', 'Micronesia', 'Montserrat', 'Niue', 'Norfolk Island', 'Pitcairn', 'Reunion', 'Saint Helena, Ascension and Tristan da Cunha', 'Saint Pierre and Miquelon', 'Saint-Barthelemy', 'Saint-Martin', 'Sint Maarten', 'South Georgia and the South Sandwich Islands', 'Tokelau', 'United States Minor Outlying Islands', 'Vatican City', 'Virgin Islands, U.S.', 'Wallis and Futuna']


# Cleaning Process

In [82]:
# =============================================================================
# 3.4 CLEANING
# =============================================================================

# --- Guardian: convert date to datetime ---
guardian_raw['date'] = pd.to_datetime(guardian_raw['date'], utc=True)

# --- Guardian: convert wordcount to numeric ---
guardian_raw['wordcount'] = pd.to_numeric(guardian_raw['wordcount'], errors='coerce')

# --- Guardian: extract year and month for time series analysis ---
guardian_raw['year'] = guardian_raw['date'].dt.year
guardian_raw['month'] = guardian_raw['date'].dt.month

# --- Guardian: drop duplicate titles ---
# Duplicate titles are the same article tagged with multiple countries.
# We keep all of them — each country-article pair is a valid observation.
# No rows dropped.

# --- Guardian: note on pillar nulls ---
# 1,101 rows have no pillar value. These are retained as pillar is
# supplementary — we use section as our primary categorical variable.

# --- ACLED: already clean — no changes needed ---

# --- Verify ---
print(guardian_raw.dtypes)
print(guardian_raw.isnull().sum())
print(guardian_raw.shape)

country                   object
date         datetime64[ns, UTC]
section                   object
wordcount                  int64
title                     object
pillar                    object
year                       int32
month                      int32
dtype: object
country         0
date            0
section         0
wordcount       0
title           0
pillar       1101
year            0
month           0
dtype: int64
(276578, 8)


# Making the clean dataset

In [41]:
# 3.6 BUILDING THE MASTER DATASET
# =============================================================================
# We build two analytical datasets:
#
# A) COUNTRY-LEVEL (one row per country)
#    → choropleth, scatter plot, matched comparisons, wordcount boxplots
#
# B) MONTHLY (one row per country per month)
#    → time series, coverage responsiveness analysis

# =============================================================================
# A) COUNTRY-LEVEL DATASET
# =============================================================================

guardian_raw['wordcount'] = pd.to_numeric(guardian_raw['wordcount'], errors='coerce')

# --- Step 1: Aggregate Guardian by country ---
guardian_agg = guardian_raw.groupby('country').agg(
    total_articles = ('title',     'count'),
    avg_wordcount  = ('wordcount', 'mean'),
    med_wordcount  = ('wordcount', 'median'),
).reset_index()

# --- Step 2: Aggregate ACLED by country (sum across 2015-2024) ---
acled_agg = acled.groupby('COUNTRY').agg(
    total_fatalities = ('FATALITIES', 'sum'),
    region_group     = ('region_group', 'first')
).reset_index().rename(columns={'COUNTRY': 'country'})

# --- Step 3: Merge Guardian and ACLED ---
# Inner join — only countries present in both datasets are kept.
# Countries with Guardian coverage but no ACLED conflict data are excluded
# as they have no severity benchmark for our central analysis.
master = guardian_agg.merge(acled_agg, on='country', how='inner')

# --- Step 4: Add section breakdown ---
# Calculate % of articles per country appearing in each Guardian section
section_breakdown = (
    guardian_raw
    .groupby(['country', 'section'])
    .size()
    .reset_index(name='count')
)
total_per_country = guardian_raw.groupby('country').size().reset_index(name='total')
section_breakdown = section_breakdown.merge(total_per_country, on='country')
section_breakdown['pct'] = section_breakdown['count'] / section_breakdown['total'] * 100

section_pivot = section_breakdown.pivot_table(
    index='country', columns='section', values='pct', fill_value=0
).reset_index()

master = master.merge(section_pivot, on='country', how='left')

# --- Step 5: Normalise coverage by fatalities ---
# Raw article counts favour countries with longer or more intense conflicts.
# Articles per 1,000 fatalities gives a fairer coverage intensity measure.
master['articles_per_1000_fatalities'] = (
    master['total_articles'] / master['total_fatalities'] * 1000
).replace([float('inf'), -float('inf')], None)

# --- Final check ---
print("=== Master Dataset ===")
print(f"Shape: {master.shape}")
print(f"Countries: {master['country'].nunique()}")
print(f"\nRegion split:\n{master['region_group'].value_counts()}")
print(f"\nSample:\n{master[['country', 'region_group', 'total_articles', 'total_fatalities', 'avg_wordcount']].head(10)}")

master.to_csv("../data/master.csv", index=False)
print("\nSaved master.csv")

# =============================================================================
# B) MONTHLY DATASET
# =============================================================================

guardian_raw['date'] = pd.to_datetime(guardian_raw['date'], utc=True)
guardian_raw['year']  = guardian_raw['date'].dt.year
guardian_raw['month'] = guardian_raw['date'].dt.month

# --- Step 1: Aggregate Guardian by country, year, month ---
guardian_monthly = guardian_raw.groupby(
    ['country', 'year', 'month']
).agg(
    article_count = ('title',     'count'),
    avg_wordcount = ('wordcount', 'mean')
).reset_index()

# --- Step 2: Distribute ACLED fatalities across months ---
# ACLED is only available at yearly granularity.
# We divide annual fatalities by 12 as a first-order approximation.
# This is a known limitation acknowledged in the conclusion.
acled_monthly = acled.copy()
acled_monthly['monthly_fatalities'] = acled_monthly['FATALITIES'] / 12
acled_monthly = acled_monthly.rename(columns={'COUNTRY': 'country', 'YEAR': 'year'})

# --- Step 3: Merge ---
monthly = guardian_monthly.merge(
    acled_monthly[['country', 'year', 'monthly_fatalities', 'region_group']],
    on=['country', 'year'],
    how='left'
)

print("=== Monthly Dataset ===")
print(f"Shape: {monthly.shape}")
print(monthly.head(10))

monthly.to_csv("../data/master_monthly.csv", index=False)
print("\nSaved master_monthly.csv")


=== Master Dataset ===
Shape: (197, 69)
Countries: 197

Region split:
region_group
Global South    129
Western          68
Name: count, dtype: int64

Sample:
               country  region_group  total_articles  total_fatalities  \
0          Afghanistan  Global South            2656            200899   
1              Albania  Global South             162                 4   
2              Algeria  Global South             130               649   
3              Andorra       Western               5                 0   
4               Angola  Global South             117               765   
5           Antarctica  Global South             434                 0   
6  Antigua and Barbuda       Western              46                 6   
7            Argentina  Global South             736               120   
8              Armenia  Global South             185               297   
9            Australia       Western           38000                 7   

   avg_wordcount  
0     98

# 4. Exploratory Data Analysis

# 5. Conclusion